File: computeWaveCondition.m
function out = computeWaveCondition(H, T, d, varargin)
% COMPUTEWAVECONDITION  Non-dimensional wave params + theory recommendation.
% Supports regular and irregular waves with Rayleigh-statistic conversion.
%
% Usage:
%   out = computeWaveCondition(H, T, d)
%   out = computeWaveCondition(Hs, Tp, d, 'waveType','irregular', ...
%                              'Hkind','Hs','plotKind','Hmax','Nwaves',1000)

    p = inputParser;
    addParameter(p,'g',9.81);
    addParameter(p,'waveType','regular');
    addParameter(p,'Hkind','Hs');
    addParameter(p,'plotKind','Hs');
    addParameter(p,'Nwaves',1000);
    parse(p,varargin{:});
    g        = p.Results.g;
    waveType = lower(p.Results.waveType);
    Hkind    = p.Results.Hkind;
    plotKind = p.Results.plotKind;
    Nwaves   = p.Results.Nwaves;

    % linear dispersion: omega^2 = g*k*tanh(k*d)
    omega = 2*pi/T;
    k = omega^2/g;
    for it = 1:200
        f  = g*k*tanh(k*d) - omega^2;
        fp = g*tanh(k*d) + g*k*d*sech(k*d)^2;
        dk = -f/fp; k = k + dk;
        if abs(dk) < 1e-12, break; end
    end
    L = 2*pi/k;

    if strcmp(waveType,'regular')
        Hs = H; Hplot = H;
    else
        Hs    = toHs(H, Hkind, Nwaves);
        Hplot = fromHs(Hs, plotKind, Nwaves);
    end

    out.waveType = waveType;
    out.Hkind    = Hkind;
    out.plotKind = plotKind;
    out.Hinput   = H;
    out.Hs       = Hs;
    out.Hplot    = Hplot;
    out.T        = T;
    out.d        = d;
    out.L        = L;
    out.k        = k;
    out.dL       = d/L;
    out.HL       = Hplot/L;
    out.Ur       = (Hplot/L)/(d/L)^3;
    out.dgT2     = d/(g*T^2);
    out.HgT2     = Hplot/(g*T^2);

    if out.dL < 0.05,      out.regime = 'Shallow water';
    elseif out.dL > 0.5,   out.regime = 'Deep water';
    else,                  out.regime = 'Intermediate depth';
    end

    x = out.dL;
    out.HLbreak = x*(0.141063*x^2 + 0.0095721*x + 0.0077829) / ...
                   (x^3 + 0.0788340*x^2 + 0.0317567*x + 0.0093407);
    out.broken = out.HL > out.HLbreak;

    if out.broken
        out.theory = 'BREAKING - no periodic theory valid';
        return
    end

    if out.Ur > 26
        out.theory = 'Cnoidal wave theory';
    else
        [HL2,HL3,HL4,HL5] = stokesOrderCurves(out.dL);
        if     out.HL <= HL2, out.theory = 'Linear (Airy) wave';
        elseif out.HL <= HL3, out.theory = 'Stokes 2nd order';
        elseif out.HL <= HL4, out.theory = 'Stokes 3rd order';
        elseif out.HL <= HL5, out.theory = 'Stokes 4th order';
        else,                 out.theory = 'Stokes 5th order';
        end
    end
end

function Hs = toHs(H, kind, N)
    switch upper(strrep(kind,'/',''))
        case {'HS','HM0','H13'}, Hs = H;
        case 'HRMS',             Hs = H * sqrt(2);
        case 'HMAX',             Hs = H / sqrt(0.5*log(N));
        otherwise, error('Unknown Hkind: %s', kind);
    end
end

function Ht = fromHs(Hs, kind, N)
    switch upper(strrep(kind,'/',''))
        case {'HS','HM0','H13'}, Ht = Hs;
        case 'HRMS',             Ht = Hs / sqrt(2);
        case 'HMAX',             Ht = Hs * sqrt(0.5*log(N));
        otherwise, error('Unknown plotKind: %s', kind);
    end
end

File: compute_wave_condition.py
import numpy as np
from stokes_order_curves import stokes_order_curves


def compute_wave_condition(H, T, d, wave_type='regular', Hkind='Hs',
                           plot_kind='Hs', N_waves=1000, g=9.81):
    omega = 2 * np.pi / T
    k = omega**2 / g
    for _ in range(200):
        f = g * k * np.tanh(k * d) - omega**2
        fp = g * np.tanh(k * d) + g * k * d / np.cosh(k * d)**2
        dk = -f / fp
        k += dk
        if abs(dk) < 1e-12:
            break
    L = 2 * np.pi / k

    if wave_type == 'regular':
        Hs = H
        Hplot = H
    else:
        Hs = _to_hs(H, Hkind, N_waves)
        Hplot = _from_hs(Hs, plot_kind, N_waves)

    dL = d / L
    HL = Hplot / L
    Ur = HL / dL**3
    dgT2 = d / (g * T**2)
    HgT2 = Hplot / (g * T**2)

    if dL < 0.05:
        regime = 'Shallow water'
    elif dL > 0.5:
        regime = 'Deep water'
    else:
        regime = 'Intermediate depth'

    HLbreak = dL * (0.141063 * dL**2 + 0.0095721 * dL + 0.0077829) / \
              (dL**3 + 0.0788340 * dL**2 + 0.0317567 * dL + 0.0093407)
    broken = HL > HLbreak

    if broken:
        theory = 'BREAKING - no periodic theory valid'
    elif Ur > 26:
        theory = 'Cnoidal wave theory'
    else:
        HL2, HL3, HL4, HL5 = stokes_order_curves(np.array([dL]))
        HL2, HL3, HL4, HL5 = HL2[0], HL3[0], HL4[0], HL5[0]
        if HL <= HL2:
            theory = 'Linear (Airy) wave'
        elif HL <= HL3:
            theory = 'Stokes 2nd order'
        elif HL <= HL4:
            theory = 'Stokes 3rd order'
        elif HL <= HL5:
            theory = 'Stokes 4th order'
        else:
            theory = 'Stokes 5th order'

    return {
        'waveType': wave_type, 'Hkind': Hkind, 'plotKind': plot_kind,
        'Hinput': H, 'Hs': Hs, 'Hplot': Hplot,
        'T': T, 'd': d, 'L': L, 'k': k,
        'dL': dL, 'HL': HL, 'Ur': Ur,
        'dgT2': dgT2, 'HgT2': HgT2,
        'regime': regime, 'HLbreak': HLbreak,
        'broken': broken, 'theory': theory,
    }


def _to_hs(H, kind, N):
    kind = kind.upper().replace('/', '')
    if kind in ('HS', 'HM0', 'H13'):
        return H
    elif kind == 'HRMS':
        return H * np.sqrt(2)
    elif kind == 'HMAX':
        return H / np.sqrt(0.5 * np.log(N))
    else:
        raise ValueError(f'Unknown Hkind: {kind}')


def _from_hs(Hs, kind, N):
    kind = kind.upper().replace('/', '')
    if kind in ('HS', 'HM0', 'H13'):
        return Hs
    elif kind == 'HRMS':
        return Hs / np.sqrt(2)
    elif kind == 'HMAX':
        return Hs * np.sqrt(0.5 * np.log(N))
    else:
        raise ValueError(f'Unknown plotKind: {kind}')
